src/factorizations/blocklanczos.jl

```diff
@@ -0,0 +1,322 @@
+
+    state.TDB[(k + 1):(k + bs_next), (k + 1):(k + bs_next)] .= Mnext
+    state.r.vec[1:bs_next] .= rₖnext.vec
+    state.norm_r = norm(rₖnext)
+    state.total_size += bs_next
+    state.r_size = bs_next
+
+    if verbosity > EACHITERATION_LEVEL
+        @info "BlockLanczos expansion to dimension $(state.total_size): subspace normres = $(normres2string(state.norm_r))"
+    end
+end
+
+function blocklanczosrecurrence(operator, V::OrthonormalBasis, Bₖ::AbstractMatrix,
+                                orth::ModifiedGramSchmidt2)
+    # Apply the operator and calculate the M. Get Xnext and Mnext.
+    bs, bs_last = size(Bₖ)
+    S = eltype(Bₖ)
+    k = length(V)
+    X = BlockVec{S}(V[(k - bs + 1):k])
+    AX = apply(operator, X)
+    M = block_inner(X, AX)
+    # Calculate the new residual. Get Rnext
+    Xlast = BlockVec{S}(V[(k - bs_last - bs + 1):(k - bs)])
+    rₖnext = compute_residual!(AX, X, M, Xlast, Bₖ')
+    ortho_basis!(rₖnext, V)
+    return rₖnext, M
+end
+
+"""
+    compute_residual!(AX::BlockVec{T,S}, X::BlockVec{T,S},
+                           M::AbstractMatrix,
+                           X_prev::BlockVec{T,S}, B_prev::AbstractMatrix) where {T,S}
+
+Computes the residual block and stores the result in `AX`.
+
+This function orthogonalizes `AX` against the two most recent basis blocks, `X` and `X_prev`.  
+Here, `AX` represents the image of the current block `X` under the action of the linear operator `A`.  
+The matrix `M` contains the inner products between `X` and `AX`, i.e., the projection of `AX` onto `X`.  
+Similarly, `B_prev` represents the projection of `AX` onto `X_prev`.
+
+The residual is computed as:
+
+```
+    AX ← AX - X * M - X_prev * B_prev
+```
+
+After this operation, `AX` is orthogonal (in the block inner product sense) to both `X` and `X_prev`.
+
+"""
+function compute_residual!(AX::BlockVec{T,S}, X::BlockVec{T,S},
+                           M::AbstractMatrix,
+                           X_prev::BlockVec{T,S}, B_prev::AbstractMatrix) where {T,S}
+    @inbounds for j in 1:length(X)
+        for i in 1:length(X)
+            AX[j] = add!!(AX[j], X[i], -M[i, j])
+        end
+        for i in 1:length(X_prev)
+            AX[j] = add!!(AX[j], X_prev[i], -B_prev[i, j])
+        end
+    end
+    return AX
+end
+
+"""
+    ortho_basis!(basis::BlockVec{T,S}, basis_sofar::OrthonormalBasis{T}) where {T,S}
+
+This function orthogonalizes the vectors in `basis` with respect to the previously orthonormalized set `basis_sofar` by using the modified Gram-Schmidt process.
+Specifically, it modifies each vector `basis[i]` by projecting out its components along the directions spanned by `basis_sofar`, i.e.,
+
+```
+    basis[i] = basis[i] - sum(j=1:length(basis_sofar)) <basis[i], basis_sofar[j]> basis_sofar[j]
+```
+
+Here,`⟨·,·⟩` denotes the inner product. The function assumes that `basis_sofar` is already orthonormal.
+"""
+function ortho_basis!(basis::BlockVec{T,S}, basis_sofar::OrthonormalBasis{T}) where {T,S}
```

Comment:
Since this function is called with the purpose of "reorthogonalising" `AX` to all of `V`, after it has already been orthogonalized to `X` and `Xprev`, I would make this explicit in the name and call this method `block_reorthogonalize!`